src/docker/remote.rs

```diff
@@ -0,0 +1,652 @@
+    create_volume_dir(engine, container, &dst_rustlib, verbose)?;
+
+    // we first copy over the toolchain file, then everything besides it.
+    // since we don't want to call docker 100x, we copy the intermediate
+    // files to a temp directory so they're cleaned up afterwards.
+    let toolchain_path = src_rustlib.join(&target.triple());
+    if toolchain_path.exists() {
+        copy_volume_files(engine, container, &toolchain_path, &dst_rustlib, verbose)?;
+    }
+
+    // now we need to copy over the host toolchain too, since it has
+    // some requirements to find std libraries, etc.
+    let image_triple = Host::X86_64UnknownLinuxGnu.triple();
+    if target.triple() != image_triple {
+        let toolchain_path = sysroot.join(&rustlib).join(image_triple);
+        copy_volume_files(engine, container, &toolchain_path, &dst_rustlib, verbose)?;
+    }
+
+    Ok(())
+}
+
+fn run_and_get_status(engine: &Engine, args: &[&str], verbose: bool) -> Result<ExitStatus> {
+    command(engine)
+        .args(args)
+        .run_and_get_status(verbose, false)
+        .map_err(Into::into)
+}
+
+pub fn volume_create(engine: &Engine, volume: &str, verbose: bool) -> Result<ExitStatus> {
+    run_and_get_status(engine, &["volume", "create", volume], verbose)
+}
+
+pub fn volume_rm(engine: &Engine, volume: &str, verbose: bool) -> Result<ExitStatus> {
+    run_and_get_status(engine, &["volume", "rm", volume], verbose)
+}
+
+pub fn volume_exists(engine: &Engine, volume: &str, verbose: bool) -> Result<bool> {
+    command(engine)
+        .args(&["volume", "inspect", volume])
+        .run_and_get_output(verbose)
+        .map(|output| output.status.success())
+        .map_err(Into::into)
+}
+
+pub fn container_stop(engine: &Engine, container: &str, verbose: bool) -> Result<ExitStatus> {
+    run_and_get_status(engine, &["stop", container], verbose)
+}
+
+pub fn container_rm(engine: &Engine, container: &str, verbose: bool) -> Result<ExitStatus> {
+    run_and_get_status(engine, &["rm", container], verbose)
+}
+
+pub fn container_state(engine: &Engine, container: &str, verbose: bool) -> Result<ContainerState> {
+    let stdout = command(engine)
+        .args(&["ps", "-a"])
+        .args(&["--filter", &format!("name={container}")])
+        .args(&["--format", "{{.State}}"])
+        .run_and_get_stdout(verbose)?;
+    ContainerState::new(stdout.trim())
+}
+
+pub fn unique_toolchain_identifier(target: &Target, sysroot: &Path) -> Result<String> {
+    // try to get the commit hash for the currently toolchain, if possible
+    // if not, get the default rustc and use the path hash for uniqueness
+    let commit_hash = if let Some(version) = rustup::rustc_version_string(sysroot)? {
+        rustc::hash_from_version_string(&version, 1)
+    } else {
+        rustc::version_meta()?.commit_hash()
+    };
+
+    let triple = target.triple();
+    let toolchain_hash = path_hash(sysroot)?;
+    Ok(format!("cross-{triple}-{toolchain_hash}-{commit_hash}"))
+}
```

Comment:
Yeah, it does currently. I can make this more general still, but it would need some memo to ensure that I know if the toolchain for the triple has been previously copied over. I guess this could also solve the issue of having `rustup` not having previously installed the cross-compile target toolchain prior to creating the persistent volume.

Reply:
> but it would need some memo to ensure that I know if the toolchain for the triple has been previously copied over.

you mean for the target? Yes that a good point is there a good way to add that to some kind of metadata in the volume? maybe having a separate volume per toolchain/target is fine then, just sad about duplicating files

Reply:
> > but it would need some memo to ensure that I know if the toolchain for the triple has been previously copied over.
> 
> you mean for the target? Yes that a good point is there a good way to add that to some kind of metadata in the volume? maybe having a separate volume per toolchain/target is fine then, just sad about duplicating files

Yeah, so currently there's a unique volume per target, mostly because I didn't want to have to deal with updating the rustlib directory every single time a new target was added, and wasn't sure if there was a good way to do so efficiently. There probably is, and it's probably a better solution which I'll look into.